### PR TITLE
Add Clipping to displaybuffer

### PR DESCRIPTION
**Pull request by nielsnl68**
# What does this implement/fix?

This per extends the displaybuffer with the option to clip parts of the screen where the changes are allowed.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
display:
  - platform: ili9341
    model: TFT 2.4
    cs_pin: 14
    dc_pin: 27
    led_pin: 32  
    reset_pin: 33

    lambda: |-
      it.print(20, 20, id(my_red), id(my_red), TextAlign::TOP_LEFT, "Hello World!");
      it.push_clipping(30, 25, 100, 40);
      it.print(20, 20, id(my_red), id(my_blue), TextAlign::TOP_LEFT, "Hello World!");
      it.pop_clipping();

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/display/display_buffer.cpp [hunk 1]
```diff
@@ -392,6 +395,43 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
 }
 #endif
 
+void DisplayBuffer::push_clipping(Rect rect) {
+  // ESP_LOGW(TAG, "set: Push new clipping");
```
Comment (jesserockz) [inline esphome/components/display/display_buffer.cpp]:
```suggestion
```

File: esphome/components/display/display_buffer.cpp [hunk 2]
```diff
@@ -392,6 +395,43 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
...
+    ESP_LOGE(TAG, "add: Clipping is not set.");
+  } else {
+    this->clipping_rectangle_.back().join(add_rect);
+  }
+}
+void DisplayBuffer::substract_clipping(Rect add_rect) {
+  if (this->clipping_rectangle_.empty()) {
+    ESP_LOGE(TAG, "add: Clipping is not set.");
+  } else {
+    // ESP_LOGW(TAG, "add: join new clipping");
```
Comment (jesserockz) [inline esphome/components/display/display_buffer.cpp]:
```suggestion
```

File: esphome/components/display/display_buffer.h [hunk 3]
```diff
@@ -374,6 +468,72 @@ class DisplayBuffer {
...
+  /// amounts on both side), based on the centerpoint of the rectangle.
+  ///
+  /// \param[in]  rect:       Rectangular region before resizing
+  /// \param[in]  width:    Number of pixels to expand the width (if positive)
+  ///                          of contract the width (if negative)
+  /// \param[in]  height:    Number of pixels to expand the height (if positive)
+  ///                          of contract the height (if negative)
+  ///
+  /// \return new rect with resized dimensions
+  ///
```
Comment (jesserockz) [inline esphome/components/display/display_buffer.h]: ???

File: esphome/components/display/display_buffer.h [hunk 4]
```diff
@@ -100,6 +101,98 @@ enum DisplayRotation {
...
+    } else {
+      return ((rect.x <= this->x2()) && (rect.x2() >= this->x) && (rect.y <= this->y2()) && (rect.y2() >= this->y));
+    }
+  }
+
+  //  rect           x--------------w
+  //  this        x--------------w
+  inline void info(const std::string &prefix = "rect info:") {
+    if (this->is_set()) {
+      ESP_LOGI("Rect", "%s [%3d,%3d,%3d,%3d]", prefix.c_str(), this->x, this->y, this->w, this->h);
```
Comment (jesserockz) [inline esphome/components/display/display_buffer.h]: Log calls need to be in the cpp file

File: esphome/components/display/display_buffer.h [hunk 5]
```diff
@@ -4,7 +4,8 @@
 #include "esphome/core/defines.h"
 #include "esphome/core/automation.h"
 #include "display_color_utils.h"
-
+#include "esphome/core/helpers.h"
+#include "esphome/core/log.h"
```
Comment (jesserockz) [inline esphome/components/display/display_buffer.h]: This needs to go in the cpp file

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/display/display_buffer.cpp (hunk 1 above)
Comment (nielsnl68) [reply esphome/components/display/display_buffer.cpp]: I have moved the function code to the cpp file now.

File: esphome/components/display/display_buffer.h (hunk 3 above)
Comment (nielsnl68) [reply esphome/components/display/display_buffer.h]: thanks, i have removed it

File: esphome/components/display/display_buffer.h (hunk 4 above)
Comment (nielsnl68) [reply esphome/components/display/display_buffer.h]: I moved the functions now to the cpp file

File: esphome/components/display/display_buffer.h (hunk 5 above)
Comment (nielsnl68) [reply esphome/components/display/display_buffer.h]: done

**Review by jesserockz — APPROVED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**